src/simulators/statevector/qv_avx2.cpp

```diff
@@ -1045,12 +1102,150 @@ template Avx apply_matrix_avx<double>(double*,
                                       const size_t qregs_size,
                                       const double* mat,
                                       const size_t omp_threads);
+
 template Avx apply_matrix_avx<float>(float* data,
                                      const uint64_t data_size,
                                      const uint64_t* qregs,
                                      const size_t qregs_size,
                                      const float* mat,
                                      const size_t omp_threads);
 
+template <>
+Avx apply_diagonal_matrix_avx<double>(double* qv_data_,
+                                      const uint64_t data_size,
+                                      const uint64_t* qregs,
+                                      const size_t qregs_size,
+                                      const double* vec_,
+                                      const size_t omp_threads) {
+
+  std::complex<double>* qv_data = reinterpret_cast<std::complex<double>*>(qv_data_);
+  const std::complex<double>* input_vec = reinterpret_cast<const std::complex<double>*>(vec_);
+
+  std::complex<double>** inputs;
+  inputs = reinterpret_cast<std::complex<double>**>(malloc(sizeof(std::complex<double*>) * omp_threads));
+
+#pragma omp parallel for if (omp_threads > 1) num_threads(omp_threads)
+  for (int i = 0; i < omp_threads; ++i) {
+#if !defined(_WIN64) && !defined(_WIN32)
+    void* data;
+    posix_memalign(&data, 64, sizeof(std::complex<double>) * 2);
+    inputs[omp_get_thread_num()] = reinterpret_cast<std::complex<double>*>(data);
```

Comment:
I think this is the part that needs a safe option for no OpenMP, something like (just guessing here for the non-omp case):

```c++
#ifdef _OPENMP
size_t index = omp_get_gread_num()
#else
size_t index = 0
#endif

inputs[index] = reinterpret_cast<std::complex<double>*>(data);
```

Similarly for other parts of the code that use `omp_get_thread_num`